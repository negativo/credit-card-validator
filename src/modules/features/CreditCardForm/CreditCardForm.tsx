import React, { useState } from 'react';
import { FormContainer, TextInput, Button } from '../../UI/components';

import { isCardValid, getCardBrand } from '../../utils';

const getMaskedInput = (value: string) => {
  const cardBrand = getCardBrand(value);
  if (cardBrand && cardBrand.name === 'Amex') {
    return value
      .replace(/[^\d0-9]/g, '')
      .replace(/(.{4})/, '$1 ')
      .replace(/(.{12})/, '$1 ')
      .trim();
  }
  return value
    .replace(/[^\d0-9]/g, '')
    .replace(/(.{4})/g, '$1 ')
    .trim();
};
export const CreditCardForm: React.StatelessComponent = ({ children }) => {
  const [cardNumber, setCardNumber] = useState('');
  const cardBrand = getCardBrand(cardNumber);
  return (
    <FormContainer>
      <TextInput
        onChange={(e: React.FormEvent<HTMLInputElement>) => {
          // Allow only numbers and Insert space every 4 chars
          if (e.currentTarget.value.length > 19) {
            return false;
          }
          const value = getMaskedInput(e.currentTarget.value);

          setCardNumber(value);
        }}
        label={`Card Number${cardBrand ? ': ' + cardBrand.name : ''}`}
        isValid={cardNumber ? isCardValid(cardNumber) : false}
        value={cardNumber}
      />

      <p />
      <Button disabled={!isCardValid(cardNumber)}>Submit</Button>
    </FormContainer>
  );
};
