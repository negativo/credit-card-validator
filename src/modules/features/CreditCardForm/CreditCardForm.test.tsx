import React from 'react';
import { CreditCardForm } from './CreditCardForm';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(<CreditCardForm />).toJSON();
  expect(tree).toMatchSnapshot();
});
