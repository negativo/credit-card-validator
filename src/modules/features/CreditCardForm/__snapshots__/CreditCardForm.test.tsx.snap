// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-AxhCb oChfx"
>
  <div
    className="sc-AxirZ fSNuUs"
  >
    <div
      className="sc-AxjAm cTPVRV"
    >
      Card Number
    </div>
    <input
      className="sc-AxiKw ckEfbS"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <p />
  <div
    className="sc-AxhUy cRhGhy"
    disabled={true}
  >
    Submit
  </div>
</div>
`;
